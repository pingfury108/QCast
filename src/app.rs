use async_trait::async_trait;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::Queue,
    boot::{create_app, BootResult, StartMode},
    config::Config,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    task::Tasks,
    Result,
};
use migration::Migrator;
use std::path::Path;

#[allow(unused_imports)]
use crate::{controllers, initializers, models::_entities::users, tasks, workers};

pub struct App;
#[async_trait]
impl Hooks for App {
    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA")
                .or(option_env!("GITHUB_SHA"))
                .unwrap_or("dev")
        )
    }

    async fn boot(
        mode: StartMode,
        environment: &Environment,
        config: Config,
    ) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment, config).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .add_route(controllers::books::routes())
            .add_route(controllers::chapters::routes())
            .add_route(controllers::medias::routes())
            .add_route(controllers::public::routes())
            .add_route(controllers::auth::routes())
            .add_route(controllers::dashboard::routes())
            // 后台管理路由
            .add_route(controllers::admin::users::routes())
            .add_route(controllers::admin::groups::routes())
    }
    async fn connect_workers(_ctx: &AppContext, _queue: &Queue) -> Result<()> {
        // 暂无 worker
        Ok(())
    }

    #[allow(unused_variables)]
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::create_superadmin::CreateSuperadmin);
        // tasks-inject (do not remove)
    }
    async fn truncate(ctx: &AppContext) -> Result<()> {
        truncate_table(&ctx.db, users::Entity).await?;
        Ok(())
    }
    async fn seed(ctx: &AppContext, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(&ctx.db, &base.join("users.yaml").display().to_string())
            .await?;
        Ok(())
    }
}
