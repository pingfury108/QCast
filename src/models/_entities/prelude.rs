//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub use super::books::Entity as Books;
pub use super::chapters::Entity as Chapters;
pub use super::medias::Entity as Medias;
pub use super::site_settings::Entity as SiteSettings;
pub use super::user_group_members::Entity as UserGroupMembers;
pub use super::user_groups::Entity as UserGroups;
pub use super::users::Entity as Users;
