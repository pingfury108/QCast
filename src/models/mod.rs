pub mod _entities;
pub mod books;
pub mod chapters;
pub mod medias;
pub mod user_group_members;
pub mod user_groups;
pub mod users;
