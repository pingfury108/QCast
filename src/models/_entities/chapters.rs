//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "chapters")]
pub struct Model {
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    #[sea_orm(primary_key)]
    pub id: i32,
    pub title: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    pub sort_order: Option<i32>,
    pub book_id: i32,
    pub parent_id: Option<i32>,
    pub level: Option<i32>,
    #[sea_orm(column_type = "Text", nullable)]
    pub path: Option<String>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::books::Entity",
        from = "Column::BookId",
        to = "super::books::Column::Id",
        on_update = "Cascade",
        on_delete = "Cascade"
    )]
    Books,
    #[sea_orm(
        belongs_to = "super::chapters::Entity",
        from = "Column::ParentId",
        to = "super::chapters::Column::Id",
        on_update = "Cascade",
        on_delete = "SetNull"
    )]
    Parent,
    #[sea_orm(has_many = "super::chapters::Entity")]
    Children,
    #[sea_orm(has_many = "super::medias::Entity")]
    Medias,
}

impl Related<super::books::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Books.def()
    }
}

impl Related<super::medias::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Medias.def()
    }
}

impl Related<Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Children.def()
    }
}
