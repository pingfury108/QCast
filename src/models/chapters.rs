pub use super::_entities::books;
pub use super::_entities::chapters::{ActiveModel, Column, Entity, Model};
use sea_orm::entity::prelude::*;
use sea_orm::IntoActiveModel;
use sea_orm::QueryOrder;
use sea_orm::QuerySelect;
use sea_orm::Set;
use sea_orm::TransactionTrait;
pub type Chapters = Entity;

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    async fn before_save<C>(self, _db: &C, insert: bool) -> std::result::Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        let mut this = self;

        if insert {
            // 设置默认值
            if this.sort_order.is_unchanged() {
                this.sort_order = sea_orm::ActiveValue::Set(Some(0));
            }
        }

        if !insert && this.updated_at.is_unchanged() {
            this.updated_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        Ok(this)
    }
}

// implement your read-oriented logic here
impl Model {
    /// 获取书籍的所有章节
    pub async fn find_by_book(db: &DatabaseConnection, book_id: i32) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::BookId.eq(book_id))
            .order_by_asc(Column::SortOrder)
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 获取带媒体计数的章节
    pub async fn find_by_book_with_media_count(db: &DatabaseConnection, book_id: i32) -> Result<Vec<(Model, i64)>, DbErr> {
        use crate::models::_entities::medias;
        use sea_orm::JoinType;

        let chapters = Entity::find()
            .filter(Column::BookId.eq(book_id))
            .order_by_asc(Column::SortOrder)
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await?;

        let mut result = Vec::new();
        for chapter in chapters {
            let media_count = medias::Entity::find()
                .filter(medias::Column::ChapterId.eq(chapter.id))
                .count(db)
                .await?;
            result.push((chapter, media_count as i64));
        }

        Ok(result)
    }

    /// 搜索章节
    pub async fn search(
        db: &DatabaseConnection,
        book_id: i32,
        query: &str,
    ) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::BookId.eq(book_id))
            .filter(
                Column::Title
                    .contains(query)
                    .or(Column::Description.contains(query)),
            )
            .order_by_asc(Column::SortOrder)
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 获取下一个排序号
    pub async fn get_next_sort_order(db: &DatabaseConnection, book_id: i32) -> Result<i32, DbErr> {
        let max_order = Entity::find()
            .filter(Column::BookId.eq(book_id))
            .select_only()
            .column(Column::SortOrder)
            .order_by_desc(Column::SortOrder)
            .into_tuple::<Option<i32>>()
            .one(db)
            .await?;

        Ok(max_order.flatten().unwrap_or(0) + 1)
    }

    /// 重新排序书籍的所有章节
    pub async fn reorder_all(
        db: &DatabaseConnection,
        book_id: i32,
        chapter_ids: &[i32],
    ) -> Result<(), DbErr> {
        let txn = db.begin().await?;

        for (new_order, &chapter_id) in chapter_ids.iter().enumerate() {
            if let Some(chapter) = Entity::find_by_id(chapter_id).one(&txn).await? {
                if chapter.book_id != book_id {
                    txn.rollback().await?;
                    return Err(DbErr::Custom(
                        "Chapter does not belong to this book".to_string(),
                    ));
                }

                let mut active_chapter = chapter.into_active_model();
                active_chapter.sort_order = Set(Some((new_order + 1) as i32));
                active_chapter.update(&txn).await?;
            } else {
                txn.rollback().await?;
                return Err(DbErr::Custom("Chapter not found".to_string()));
            }
        }

        txn.commit().await?;
        Ok(())
    }

    /// 向上移动章节（减小 sort_order）
    pub async fn move_up(db: &DatabaseConnection, chapter_id: i32) -> Result<bool, DbErr> {
        let chapter = Entity::find_by_id(chapter_id).one(db).await?;
        if let Some(chapter) = chapter {
            let current_order = chapter.sort_order.unwrap_or(0);

            // 找到当前章节前一个章节
            let prev_chapter = Entity::find()
                .filter(Column::BookId.eq(chapter.book_id))
                .filter(Column::SortOrder.lt(current_order))
                .order_by_desc(Column::SortOrder)
                .one(db)
                .await?;

            if let Some(prev_chapter) = prev_chapter {
                let txn = db.begin().await?;

                // 保存原来的排序号
                let prev_order = prev_chapter.sort_order;

                // 交换排序号
                let mut active_current = chapter.into_active_model();
                let mut active_prev = prev_chapter.into_active_model();

                active_current.sort_order = Set(prev_order);
                active_prev.sort_order = Set(Some(current_order));

                active_current.update(&txn).await?;
                active_prev.update(&txn).await?;

                txn.commit().await?;
                Ok(true)
            } else {
                Ok(false) // 已经是第一个章节
            }
        } else {
            Err(DbErr::Custom("Chapter not found".to_string()))
        }
    }

    /// 向下移动章节（增加 sort_order）
    pub async fn move_down(db: &DatabaseConnection, chapter_id: i32) -> Result<bool, DbErr> {
        let chapter = Entity::find_by_id(chapter_id).one(db).await?;
        if let Some(chapter) = chapter {
            let current_order = chapter.sort_order.unwrap_or(0);

            // 找到当前章节后一个章节
            let next_chapter = Entity::find()
                .filter(Column::BookId.eq(chapter.book_id))
                .filter(Column::SortOrder.gt(current_order))
                .order_by_asc(Column::SortOrder)
                .one(db)
                .await?;

            if let Some(next_chapter) = next_chapter {
                let txn = db.begin().await?;

                // 保存原来的排序号
                let next_order = next_chapter.sort_order;

                // 交换排序号
                let mut active_current = chapter.into_active_model();
                let mut active_next = next_chapter.into_active_model();

                active_current.sort_order = Set(next_order);
                active_next.sort_order = Set(Some(current_order));

                active_current.update(&txn).await?;
                active_next.update(&txn).await?;

                txn.commit().await?;
                Ok(true)
            } else {
                Ok(false) // 已经是最后一个章节
            }
        } else {
            Err(DbErr::Custom("Chapter not found".to_string()))
        }
    }
}

// implement your write-oriented logic here
impl ActiveModel {
    /// 创建新章节时自动设置排序号
    pub async fn create_with_order(
        db: &DatabaseConnection,
        book_id: i32,
        title: String,
        description: Option<String>,
    ) -> Result<Self, DbErr> {
        let next_order = Model::get_next_sort_order(db, book_id).await?;

        Ok(Self {
            book_id: Set(book_id),
            title: Set(title),
            description: Set(description),
            sort_order: Set(Some(next_order)),
            ..Default::default()
        })
    }
}

// implement your custom finders, selectors oriented logic here
impl Entity {}
