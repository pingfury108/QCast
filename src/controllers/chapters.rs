#![allow(clippy::missing_errors_doc)]
#![allow(clippy::unnecessary_struct_initialization)]
#![allow(clippy::unused_async)]
use axum::debug_handler;
use axum::routing::method_routing::delete as axum_delete;
use loco_rs::prelude::*;
use serde::{Deserialize, Serialize};

use crate::models::_entities::chapters::{ActiveModel, Entity, Model};
use crate::models::books;
use crate::models::users;
use crate::views::chapters::ChapterResponse;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateParams {
    pub title: String,
    pub description: Option<String>,
    pub sort_order: Option<i32>,
    pub parent_id: Option<i32>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateParams {
    pub title: Option<String>,
    pub description: Option<String>,
    pub sort_order: Option<i32>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ReorderParams {
    pub sort_order: i32,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct BatchReorderParams {
    pub chapter_ids: Vec<i32>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateChildParams {
    pub title: String,
    pub description: Option<String>,
    pub sort_order: Option<i32>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct MoveChapterParams {
    pub new_parent_id: Option<i32>,
    pub new_sort_order: Option<i32>,
}

async fn load_item(ctx: &AppContext, id: i32, user_id: i32) -> Result<Model> {
    let item = Entity::find_by_id(id).one(&ctx.db).await?;

    if let Some(ref chapter) = item {
        // 验证书籍是否属于当前用户
        let book = books::Entity::find_by_id(chapter.book_id)
            .filter(books::Column::UserId.eq(user_id))
            .one(&ctx.db)
            .await?;

        if book.is_none() {
            return Err(Error::NotFound);
        }
    }

    item.ok_or_else(|| Error::NotFound)
}


/// 获取书籍的章节列表
#[debug_handler]
pub async fn list(
    auth: auth::JWT,
    Path(book_id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    // 验证用户是否有权限访问该书籍
    let _book = books::Entity::find_by_id(book_id)
        .filter(books::Column::UserId.eq(user.id))
        .one(&ctx.db)
        .await?
        .ok_or_else(|| Error::NotFound)?;

    let chapters_with_media = Model::find_by_book_with_media_count(&ctx.db, book_id).await?;
    let responses: Vec<ChapterResponse> = chapters_with_media.into_iter().map(|(chapter, media_count)| {
        ChapterResponse::from_with_media_count(chapter, media_count)
    }).collect();

    format::json(responses)
}

/// 创建章节
#[debug_handler]
pub async fn create(
    auth: auth::JWT,
    Path(book_id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<CreateParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    // 验证用户是否有权限访问该书籍
    let _book = books::Entity::find_by_id(book_id)
        .filter(books::Column::UserId.eq(user.id))
        .one(&ctx.db)
        .await?
        .ok_or_else(|| Error::NotFound)?;

    let item = if let Some(parent_chapter_id) = params.parent_id {
        // 创建子章节
        if let Some(sort_order) = params.sort_order {
            // 使用指定的排序号
            ActiveModel {
                book_id: Set(book_id),
                parent_id: Set(Some(parent_chapter_id)),
                title: Set(params.title),
                description: Set(params.description),
                sort_order: Set(Some(sort_order)),
                ..Default::default()
            }
            .insert(&ctx.db)
            .await?
        } else {
            // 自动获取同级下一个排序号
            let item = ActiveModel::create_child(&ctx.db, book_id, parent_chapter_id, params.title, params.description)
                .await?;
            item.insert(&ctx.db).await?
        }
    } else if let Some(sort_order) = params.sort_order {
        // 创建顶级章节，使用指定的排序号
        ActiveModel {
            book_id: Set(book_id),
            title: Set(params.title),
            description: Set(params.description),
            sort_order: Set(Some(sort_order)),
            ..Default::default()
        }
        .insert(&ctx.db)
        .await?
    } else {
        // 创建顶级章节，自动获取下一个排序号
        let item =
            ActiveModel::create_with_order(&ctx.db, book_id, params.title, params.description)
                .await?;
        item.insert(&ctx.db).await?
    };

    // 更新层级和路径信息
    crate::models::chapters::Model::update_level_and_path(&ctx.db, item.id, params.parent_id).await?;

    // 重新加载更新的数据
    let updated_item = crate::models::_entities::chapters::Entity::find_by_id(item.id)
        .one(&ctx.db)
        .await?
        .ok_or(Error::NotFound)?;

    format::json(ChapterResponse::from(updated_item))
}

/// 获取章节详情
#[debug_handler]
pub async fn show(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    format::json(ChapterResponse::from(item))
}

/// 更新章节
#[debug_handler]
pub async fn update(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
    Json(params): Json<UpdateParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    let mut item = item.into_active_model();

    if let Some(title) = params.title {
        item.title = Set(title);
    }
    if let Some(description) = params.description {
        item.description = Set(Some(description));
    }
    if let Some(sort_order) = params.sort_order {
        item.sort_order = Set(Some(sort_order));
    }

    let item = item.update(&ctx.db).await?;

    format::json(ChapterResponse::from(item))
}

/// 删除章节
#[debug_handler]
pub async fn delete(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    item.delete(&ctx.db).await?;
    format::empty()
}

/// 调整章节顺序
#[debug_handler]
pub async fn reorder(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
    Json(params): Json<ReorderParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    let mut item = item.into_active_model();
    item.sort_order = Set(Some(params.sort_order));
    let item = item.update(&ctx.db).await?;

    format::json(ChapterResponse::from(item))
}

/// 批量重排序章节
#[debug_handler]
pub async fn batch_reorder(
    auth: auth::JWT,
    Path(book_id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<BatchReorderParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    // 验证用户是否有权限访问该书籍
    let _book = books::Entity::find_by_id(book_id)
        .filter(books::Column::UserId.eq(user.id))
        .one(&ctx.db)
        .await?
        .ok_or_else(|| Error::NotFound)?;

    Model::reorder_all(&ctx.db, book_id, &params.chapter_ids).await?;

    let chapters = Model::find_by_book(&ctx.db, book_id).await?;
    let responses: Vec<ChapterResponse> = chapters.into_iter().map(ChapterResponse::from).collect();

    format::json(responses)
}

/// 章节上移
#[debug_handler]
pub async fn move_up(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let _item = load_item(&ctx, id, user.id).await?;

    let moved = Model::move_up(&ctx.db, id).await?;

    if !moved {
        return Err(Error::Message("Chapter is already at the top".to_string()));
    }

    let updated_item = Entity::find_by_id(id).one(&ctx.db).await?.unwrap();
    format::json(ChapterResponse::from(updated_item))
}

/// 章节下移
#[debug_handler]
pub async fn move_down(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let _item = load_item(&ctx, id, user.id).await?;

    let moved = Model::move_down(&ctx.db, id).await?;

    if !moved {
        return Err(Error::Message(
            "Chapter is already at the bottom".to_string(),
        ));
    }

    let updated_item = Entity::find_by_id(id).one(&ctx.db).await?.unwrap();
    format::json(ChapterResponse::from(updated_item))
}

/// 获取书籍的章节树状结构
#[debug_handler]
pub async fn tree(
    auth: auth::JWT,
    Path(book_id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    // 验证用户是否有权限访问该书籍
    let _book = books::Entity::find_by_id(book_id)
        .filter(books::Column::UserId.eq(user.id))
        .one(&ctx.db)
        .await?
        .ok_or_else(|| Error::NotFound)?;

    let chapter_tree = Model::get_book_tree(&ctx.db, book_id).await?;
    format::json(chapter_tree)
}

/// 获取章节的扁平列表（包含层级信息）
#[debug_handler]
pub async fn flat_list(
    auth: auth::JWT,
    Path(book_id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    // 验证用户是否有权限访问该书籍
    let _book = books::Entity::find_by_id(book_id)
        .filter(books::Column::UserId.eq(user.id))
        .one(&ctx.db)
        .await?
        .ok_or_else(|| Error::NotFound)?;

    let flat_list = Model::get_flat_list_with_level(&ctx.db, book_id).await?;
    format::json(flat_list)
}

/// 获取章节的子章节
#[debug_handler]
pub async fn children(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let _chapter = load_item(&ctx, id, user.id).await?;

    let children = Model::find_children(&ctx.db, id).await?;
    let responses: Vec<ChapterResponse> = children.into_iter().map(ChapterResponse::from).collect();

    format::json(responses)
}

/// 创建子章节
#[debug_handler]
pub async fn create_child(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
    Json(params): Json<CreateChildParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let parent_chapter = load_item(&ctx, id, user.id).await?;

    let item = if let Some(sort_order) = params.sort_order {
        // 使用指定的排序号
        ActiveModel {
            book_id: Set(parent_chapter.book_id),
            parent_id: Set(Some(id)),
            title: Set(params.title),
            description: Set(params.description),
            sort_order: Set(Some(sort_order)),
            ..Default::default()
        }
        .insert(&ctx.db)
        .await?
    } else {
        // 自动获取同级下一个排序号
        let item = ActiveModel::create_child(&ctx.db, parent_chapter.book_id, id, params.title, params.description)
            .await?;
        item.insert(&ctx.db).await?
    };

    // 更新层级和路径信息
    Model::update_level_and_path(&ctx.db, item.id, Some(id)).await?;

    // 重新加载更新的数据
    let updated_item = crate::models::_entities::chapters::Entity::find_by_id(item.id)
        .one(&ctx.db)
        .await?
        .ok_or(Error::NotFound)?;

    format::json(ChapterResponse::from(updated_item))
}

/// 移动章节到新的父级
#[debug_handler]
pub async fn move_chapter(
    auth: auth::JWT,
    Path((_book_id, id)): Path<(i32, i32)>,
    State(ctx): State<AppContext>,
    Json(params): Json<MoveChapterParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let _chapter = load_item(&ctx, id, user.id).await?;

    // 如果指定了新的父级，验证父级章节是否存在且属于同一本书
    if let Some(new_parent_id) = params.new_parent_id {
        let parent_chapter = Entity::find_by_id(new_parent_id).one(&ctx.db).await?;
        if let Some(parent_chapter) = parent_chapter {
            // 验证书籍是否属于当前用户
            let _book = books::Entity::find_by_id(parent_chapter.book_id)
                .filter(books::Column::UserId.eq(user.id))
                .one(&ctx.db)
                .await?
                .ok_or_else(|| Error::NotFound)?;
        } else {
            return Err(Error::NotFound);
        }
    }

      if let Err(e) = Model::move_to_parent(&ctx.db, id, params.new_parent_id, params.new_sort_order).await {
        // 处理循环引用错误
        match e {
            sea_orm::DbErr::Custom(ref msg) if msg.contains("cycle") => {
                return Err(Error::BadRequest("Moving chapter would create a cycle".to_string()));
            }
            _ => return Err(e.into()),
        }
    }

    let updated_item = Entity::find_by_id(id).one(&ctx.db).await?.unwrap();
    format::json(ChapterResponse::from(updated_item))
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("/api/books")
        .add("/{book_id}/chapters", get(list))
        .add("/{book_id}/chapters", post(create))
        .add("/{book_id}/chapters/batch-reorder", post(batch_reorder))
        .add("/{book_id}/chapters/tree", get(tree))
        .add("/{book_id}/chapters/flat", get(flat_list))
        .add("/{book_id}/chapters/{id}/reorder", post(reorder))
        .add("/{book_id}/chapters/{id}/move-up", post(move_up))
        .add("/{book_id}/chapters/{id}/move-down", post(move_down))
        .add("/{book_id}/chapters/{id}/children", get(children))
        .add("/{book_id}/chapters/{id}/children", post(create_child))
        .add("/{book_id}/chapters/{id}/move", post(move_chapter))
        .add("/{book_id}/chapters/{id}", get(show))
        .add("/{book_id}/chapters/{id}", put(update))
        .add("/{book_id}/chapters/{id}", patch(update))
        .add("/{book_id}/chapters/{id}", axum_delete(delete))
}
