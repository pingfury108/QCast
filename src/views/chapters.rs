use crate::models::_entities::chapters::Model;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct ChapterResponse {
    pub id: i32,
    pub title: String,
    pub description: Option<String>,
    pub sort_order: Option<i32>,
    pub book_id: i32,
    pub parent_id: Option<i32>,
    pub level: Option<i32>,
    pub path: Option<String>,
    pub media_count: i64,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

impl From<Model> for ChapterResponse {
    fn from(chapter: Model) -> Self {
        Self {
            id: chapter.id,
            title: chapter.title,
            description: chapter.description,
            sort_order: chapter.sort_order,
            book_id: chapter.book_id,
            parent_id: chapter.parent_id,
            level: chapter.level,
            path: chapter.path,
            media_count: 0, // 默认值，需要从其他地方计算
            created_at: chapter.created_at.naive_utc().and_utc(),
            updated_at: chapter.updated_at.naive_utc().and_utc(),
        }
    }
}

impl From<(Model, i64)> for ChapterResponse {
    fn from((chapter, media_count): (Model, i64)) -> Self {
        Self {
            id: chapter.id,
            title: chapter.title,
            description: chapter.description,
            sort_order: chapter.sort_order,
            book_id: chapter.book_id,
            parent_id: chapter.parent_id,
            level: chapter.level,
            path: chapter.path,
            media_count,
            created_at: chapter.created_at.naive_utc().and_utc(),
            updated_at: chapter.updated_at.naive_utc().and_utc(),
        }
    }
}

impl ChapterResponse {
    /// 从带媒体计数的章节创建响应
    #[must_use]
    pub fn from_with_media_count(chapter: Model, media_count: i64) -> Self {
        Self {
            id: chapter.id,
            title: chapter.title,
            description: chapter.description,
            sort_order: chapter.sort_order,
            book_id: chapter.book_id,
            parent_id: chapter.parent_id,
            level: chapter.level,
            path: chapter.path,
            media_count,
            created_at: chapter.created_at.naive_utc().and_utc(),
            updated_at: chapter.updated_at.naive_utc().and_utc(),
        }
    }

    /// 从章节树结构创建响应
    #[must_use]
    pub fn from_chapter_tree(chapter_tree: crate::models::chapters::ChapterTree) -> Self {
        Self {
            id: chapter_tree.id,
            title: chapter_tree.title,
            description: chapter_tree.description,
            sort_order: chapter_tree.sort_order,
            book_id: chapter_tree.book_id,
            parent_id: chapter_tree.parent_id,
            level: chapter_tree.level,
            path: chapter_tree.path,
            media_count: chapter_tree.media_count,
            created_at: chapter_tree.created_at,
            updated_at: chapter_tree.updated_at,
        }
    }
}
