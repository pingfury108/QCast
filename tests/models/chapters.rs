use insta::{assert_debug_snapshot, with_settings};
use loco_rs::testing::prelude::*;
use qcast::app::App;
use qcast::models::_entities::chapters::ActiveModel;
use qcast::models::chapters;
use sea_orm::{ActiveModelTrait, Set};
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_search_chapters_with_parents() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建一个测试书籍
    let book = qcast::models::_entities::books::ActiveModel {
        title: Set("测试书籍".to_string()),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 创建章节层级结构
    let root_chapter = ActiveModel {
        title: Set("根章节".to_string()),
        book_id: Set(book.id),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let child_chapter = ActiveModel {
        title: Set("子章节".to_string()),
        book_id: Set(book.id),
        parent_id: Set(Some(root_chapter.id)),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let grandchild_chapter = ActiveModel {
        title: Set("孙子章节目标".to_string()),
        description: Set(Some("这是目标章节".to_string())),
        book_id: Set(book.id),
        parent_id: Set(Some(child_chapter.id)),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 创建一个无关的章节
    let _other_chapter = ActiveModel {
        title: Set("无关章节".to_string()),
        book_id: Set(book.id),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 使用 search_with_parents 搜索包含"目标"的章节
    let search_results =
        chapters::Model::search_with_parents(&boot.app_context.db, book.id, "目标")
            .await
            .unwrap();

    // 应该返回孙子章节、子章节和根章节（3个）
    assert_eq!(search_results.len(), 3);

    // 验证包含所有祖先
    let ids: Vec<i32> = search_results.iter().map(|c| c.id).collect();
    assert!(ids.contains(&root_chapter.id));
    assert!(ids.contains(&child_chapter.id));
    assert!(ids.contains(&grandchild_chapter.id));

    // 验证不包含无关章节
    assert!(!ids.contains(&_other_chapter.id));

    with_settings!({
        filters => cleanup_chapter_model()
    }, {
        assert_debug_snapshot!(search_results);
    });
}

fn cleanup_chapter_model() -> Vec<(&'static str, &'static str)> {
    vec![
        (r"id: \d+,", "id: ID"),
        (r"book_id: \d+,", "book_id: BOOK_ID"),
        (r"parent_id: Some\(\d+\)", "parent_id: Some(PARENT_ID)"),
        (
            r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?\+\d{2}:\d{2}",
            "DATE",
        ),
    ]
}
