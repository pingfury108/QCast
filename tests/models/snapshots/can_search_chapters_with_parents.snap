---
source: tests/models/chapters.rs
expression: search_results
---
[
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        title: "根章节",
        description: None,
        sort_order: Some(
            1,
        ),
        book_id: ID
        parent_id: None,
        level: None,
        path: None,
    },
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        title: "子章节",
        description: None,
        sort_order: Some(
            1,
        ),
        book_id: ID
        parent_id: Some(
            1,
        ),
        level: None,
        path: None,
    },
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        title: "孙子章节目标",
        description: Some(
            "这是目标章节",
        ),
        sort_order: Some(
            1,
        ),
        book_id: ID
        parent_id: Some(
            2,
        ),
        level: None,
        path: None,
    },
]
